### ISSUE TITLE
[System.Text.Json] Add ability to serialize Comments on Properties

### ISSUE THREAD

**Issue by nikeee** [dotnet/runtime#35251]
System.Text.Json already supports comments during deserialization.

JSON config files of VSCode and TypeScript and other programs use comments on their properties to add some in-line documentation.

For programs that use JSON as a config format, it would be nice if they could just annotate their classes and have these annotations as comments in the serialized JSON. This would be useful to generate a default/initial configuration file.

For example, consider:
```csharp
class Root
{
    [JsonPropertyName("test")]
    public float test = 0.42f;

    [JsonPropertyName("growIncrement")]
    [JsonComment("Lorem Ipsum")]
    public float GrowIncrement = 0.42f;

    public string test2 = "foo";
}
```
Could serialize to something like this:
```jsonc
{
    "test": 0.42,
    // Lorem Ipsum
    "growIncrement": 0.42,
    "test2": "foo"
}
```
or this:
```jsonc
{
    "test": 0.42,
    "growIncrement": 0.42, // Lorem Ipsum
    "test2": "foo"
}
```
Of course, multi-line comments could be used instead of single-line comments (just mimicking VSCode's config here).

If not baked directly into the API, it would be nice if we had the ability to add this comment serialization as a separate library.

There is some need for the feature according to this SO question: https://stackoverflow.com/questions/9495718 and https://github.com/JamesNK/Newtonsoft.Json/issues/1211

**Comment by FyiurAmron** [comment]
No. That's not "JSON". It's a JSON-like format. As long as JSON has official spec, and something doesn't conform to that spec, it's not "JSON" by any means. Use JSON5, YAML, XML, or whatever you like instead. Don't expect the only stable and sensible data format in the wild to break just because you can't understand how to use it.

**Comment by nikeee** [comment]
This is a valid feature request that (according to the votes) other users also need.

> No. That's not "JSON". It's a JSON-like format

That's right. As long as it's opt in (and that's what I am proposing here), the default behavior conforms with "the" standard. The parser already has support for comments in deserialization (which is also not "JSON compliant"). This feature request just asks to extend that support.

> because you can't understand how to use it

Please be polite. I didn't to anything to you.

**Comment by FyiurAmron** [comment]
@nikeee have you ever read what are the main reasons that `System.Text.Json` ever came to this world? From what you write, no.

>This is a valid feature request that (according to the votes) other users also need.

No. The validity of a request is based on its merit, not on popular vote. If 1000 of people voted that this library should deserialize XML too, is that a reason to do it? You're playing https://en.wikipedia.org/wiki/Argumentum_ad_populum here.

Also, you got, like, 6 upvotes during half a year? That's, like, 10 a year tops? Do you know how many people, time and code are needed to support such feature in a framework/language like .NET/C#?

Think about it.

>As long as it's opt in (and that's what I am proposing here)

If that's what you're proposing, IMO it's not a valid place to do it. Fork the lib, provide the users with that experience. Don't spam here.

Current implementation (allowing *skipping*/*ignoring* those comments) was made because *too many people abuse JSON this way, and consumer has to always be more lax than the producer*.

>as of .NET Core 3.1, JsonDocument and JsonSerializer only support skipping or disallowing of comments; they do not support loading them. If you try to set JsonCommentHandling.Allow for either, you will get an exception

Also, https://github.com/dotnet/runtime/issues/30743 etc.

Thank God there are sane people still around.

>Also, notice my comment requesting this feature as a separate library if it doesn't make it into the core APIs. That would be fine, too.

Read what Eric Lippert says about such "features". I don't always agree with him, but he is adamant on this.

>Please be polite.

I am, more than I should in this particular case.

>I didn't to anything to you.

Yes, you did; you tried, intentionally and willfully, to degrade JSON as a standard, by forcing _Yet Another Stupid Discussion About Comments In JSON_ to possibly take place. It's been discussed _ad nauseam_ for the last, like, 10 years? FWIW, people like you should be put in pillory just to make an example how not to approach software engineering.

>We're all just trying to create APIs that make us developers happy and more productive.

No. Software is not about some abstract "happiness" of some "us developers". You can't define "happiness", and you can't define "us" or "we all" in this sentence". Software is about solving engineering problems. Your problem is a X/Y problem. JSON doesn't need comments. There are already JSON extensions with comments, which are backwards compatible with it. If you want your comments to be (de)serialized - use any of those formats, they already have tons of parsers/bindings for C#.

This makes your request either invalid because of scope of this repo, or because of its nonconformity with spec and caveats associated.

**Comment by FyiurAmron** [comment]
Also, we already have https://github.com/dotnet/runtime/issues/29543 , https://github.com/dotnet/runtime/issues/30608 etc. Adding _Yet Another Obscure Case To Handle_ is IMVHO not only asking for trouble, but outright demanding it.

**Comment by nikeee** [comment]
How about offering primitives, so that 3rd party libraries could implement that?

**Comment by eiriktsarpalis** [comment]
It might be possible in the future via #60518. We're considering implementing this in 7.0.0

**Comment by nikeee** [comment]
Does this have a similar API surface as Newtonsoft's JSON? If I get this right, this would not offer the ability to implement the feautre proposed here. More details in this issue: https://github.com/JamesNK/Newtonsoft.Json/issues/1211

**Comment by FyiurAmron** [comment]
@nikeee
>How about the respective primitives, so that 3rd party libraries could implement that (which I also proposed)? Or is that possible today?

No, it's not possible, since current design consists of tightly-coupled classes, with virtually no way to inject 3rd-party behaviour in the (de)serializer. It's borderline possible via reflection and assembly hacking, but I assume that's not what you want.

>Pulling in an entirely separate JSON parsing library would be counter-productive

Once again, you're barking at a wrong tree. Please, for the sake of the argument, read what I wrote at least once with actual understanding on your side. It's not a "JSON parsing library" once you ditch the RFC. You require something that is _not_ a JSON library. I'm not arguing you shouldn't do it. Do it, if it benefits you. Just be aware that using a _non-standard_ format means you'll have to support it *yourself*, and that may (or may not) be _counter-productive_.

As far as I know, support for emitting comments was omitted (even if intended at some point, at least looking at the docs & code) from `Utf8JsonWriter` - it would make the code even more complicated & byzantine than it is now. Amount of classes that would have to be modified to support it ATM is easily reaching 10 (virtually entire serializer stack), unless you want to do some dirty hacks for it.

First, `Utf8JsonWriter` emits the separator when writing the next property, not the current one - that logic would have to be changed to allow emitting comments the way you want (and it's not that trivial of a change, even if it would seem so, if you'll check the code). Second, `Utf8JsonWriter` would have to have a completely new set of `WriteEndXXX()` functions, parametrized with your optional comment (or maybe have `WriteEndComment` or something similar), and that function(s) would have to be abstracted to those `JsonConverter`s that possibly would be able to work with the data entities you want to serialize (basically, at least [this Converter](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectDefaultConverter.cs), possibly other ones) and to extract the data from the attribute. And that's just the beginning of it, with no spec from you how to e.g. handle nested comments, handle multi-line comments, how to handle comment placement in writer in "minimize" mode etc.

Currently, those classes are `internal` (and often also `sealed`) for a reason. I guess you should be able to see the reason based on the code by yourself by now. Changing it without a major rewrite is currently hard-to-impossible, and that's for a good reason: current implementation was intended to cover RFC JSON and be as performant as possible, as a core lib. I understand why it wasn't intended to be extended. Doing so without sacrificing either performance and/or lots of developer-hours is impossible in this case.

If you really think this feature is worthwhile enough to warrant rewriting a lib that took, in essence, some years to complete, feel free to make a proposal to change the design of `System.Text.Json` classes to de-internalize the code in question (essentially: rewrite it from scratch), allowing injection via interfaces instead of current tightly-coupled design, then get enough traction and developer manpower, and voila. Honestly, I would support that notion, if only for the sake of looking at the lulz that would ensue.

>I'm also open for recommendations from @FyiurAmron for libraries that support that proposed feature

I've already did that; I've already answered your question. FWIW, did you wonder why e.g. [JSON5 for C#](https://github.com/json5/json5-dotnet) was abandoned?

(Hint: yes, it's because it's a big and non-trivial piece of code, and it has to be maintained, and nobody wanted to do it)
